kernel/pipes.c

```diff
@@ -478,6 +478,16 @@ int z_impl_k_pipe_put(struct k_pipe *pipe, void *data, size_t bytes_to_write,
 
 	z_sched_wait(&pipe->lock, key, &pipe->wait_q.writers, timeout, NULL);
 
+	/*
+	 * On SMP systems, threads in the processing list may timeout before
+	 * the data has finished copying. The following spin lock/unlock pair
+	 * prevents those threads from executing further until the data copying
+	 * is complete.
```

Comment:
I'm not quite understanding how this acts as a barrier?  I'm guessing the thinking is something like "if another CPU already holds the lock, this forces us to wait until it releases it".  But that doesn't protect against the case where it gets the lock immediately after.

Basically this doesn't feel like a fix to me.  What's the exact sequencing we're trying to protect?  My guess is something needs to be inside a lock that isn't.

Reply:
Below is the sequence of events.

1. Thread A is in the process of copying a bunch of data for pipe P on core 1.
2. Meanwhile core 2 unpends thread B that was previously waiting on pipe P.
3. Core 2 schedules thread B to run, while thread A is still copying data on core 1.

Both threads A and B then want to simultaneously access thread B's pipe descriptor--A to update it, B to read it. Worse, thread B might execute, block on pipe Q and then thread A starts updating B's pipe descriptor mistakenly thinking that it is safe to do so, when in fact it is now being used by a different pipe altogether.

What we need to do is stop thread B from executing further until the thread A has finished the pipe operation, and the spin lock is the simplest and most effective method I can think of to do it.